core-blocks/audio/edit.js

```diff
@@ -36,6 +38,30 @@ class AudioEdit extends Component {
 		this.onSelectURL = this.onSelectURL.bind( this );
 	}
 
+	componentDidMount() {
+		const { attributes, noticeOperations, setAttributes } = this.props;
+		const { id, src = '' } = attributes;
+
+		if ( ! id && src.indexOf( 'blob:' ) === 0 ) {
```

Comment:
Unfortunately using this logic to decide when to upload is vulnerable to our undo problem.
If after the upload we undo, as the set of the attributes to the blob and to the final URL are two different levels we create two undo levels.
When the user undoes the upload, we will get back to the state where we have a blob URL so the upload will happen again making the undo operation useless. It is possible to see the problem by uploading an audio block, waiting for the upload to happen, pressing undo, waiting a little bit, pressing undo again, waiting, pressing undo again and checking we are blocked.

This a complex problem related to undo for which we did not decide yet the way to attack it https://github.com/WordPress/gutenberg/pull/6325#issuecomment-402242054.

I think that maybe for this cases, (the gallery has similar code in the shortcode to gallery transform) we should use the state of the component to flag that the upload/request was made and in that case, the component does not repeat the action.

Reply:
I also think storing the blobUrl as an attribute is ok. What I was suggesting is just having a flag ( e.g: uploadCompleted ) in the state so we don't repeat the upload.

Reply:
> It is possible to see the problem by uploading an audio block, waiting for the upload to happen, pressing undo, waiting a little bit, pressing undo again, waiting, pressing undo again and checking we are blocked.

@jorgefilipecosta 🤔 I'm seeing a slightly different behaviour.

I drop an audio file, the `AudioEdit` gets a blob src, and eventually obtains a proper file src.

Now, I hit Undo.
The `AudioEdit` loses the proper file src and gets back the blob one.
And that's it. No new upload starts now.

Hitting Undo again, the Audio component loses the blob src too, and it's removed altogether.
If I hit Redo now, the Audio component will get the blob src back, and it will trigger a new upload (the new audio file has an incremented ID).

I guess this happens because the autoupload only happens on `componentDidMount`, which is not triggered on the first Undo (while it's indeed triggered on Undo-Undo-Redo, because the component is unmounted on the second Undo).

In other words, I don't think an `uploadCompleted` would really help us:
- On first Undo, the upload does not start again anyway.
- On Undo-Undo-Redo, the component is unmounted and mounted again from scratch, losing the internal state, so the file would be re-uploaded anyway.

Reply:
Hi @Copons, sorry for the delay in answering this notification got lost in the middle of notifications received while AFK.
I tried to retest this but I'm having problems executing this branch 
```
Error: Minified React error #130; visit https://reactjs.org/docs/error-decoder.html?invariant=130&args[]=undefined&args[]= for the full message or use the non-minified dev environment for full errors and additional helpful warnings. 
```
I think its related to other changes made in parallel, would it be possible to rebase this branch so we can have a new look?
Thank you in advance.